File: ResourceManager/Models/Tag.cs
using ResourceManager.Models.Abstractions;
using System.Collections.Generic;

namespace ResourceManager.Models
{
    public class Tag : BaseEntity, IItem
    {
        public string Name { get; set; } = string.Empty;
        public virtual IList<PasswordTag> PasswordTags { get; set; } = new List<PasswordTag>();
        public virtual IList<BookTag> BookTags { get; set; } = new List<BookTag>();
        public virtual IList<ArticleTag> ArticleTags { get; set; } = new List<ArticleTag>();
        public virtual IList<ActivityTag> ActivityTags { get; set; } = new List<ActivityTag>();
        public virtual IList<RepositoryTag> RepositoryTags { get; set; } = new List<RepositoryTag>();
        public virtual IList<NoteTag> NoteTags { get; set; } = new List<NoteTag>();

        public int PasswordCount { get { return PasswordTags.Count; } }
        public int BookCount { get { return BookTags.Count; } }
        public int ArticleCount { get { return ArticleTags.Count; } }
        public int ActivityCount { get { return ActivityTags.Count; } }
        public int RepositoryCount { get { return RepositoryTags.Count; } }
        public int NoteCount { get { return NoteTags.Count; } }

        public override string GetCaption()
        {
            return string.Format("{0} with Name '{1}'", nameof(Tag), Name);
        }
    }
}

File: ResourceManager/Models/Tag.vb
Imports ResourceManager.Models.Abstractions
Imports System.Collections.Generic

Namespace ResourceManager.Models
    Public Class Tag
        Inherits BaseEntity
        Implements IItem
        Public Property Name As String = String.Empty Implements IItem.Name
        Public Overridable Property PasswordTags As ICollection(Of PasswordTag)
        Public Overridable Property BookTags As ICollection(Of BookTag)
        Public Overridable Property ArticleTags As ICollection(Of ArticleTag)
        Public Overridable Property ActivityTags As ICollection(Of ActivityTag)
        Public Overridable Property RepositoryTags As ICollection(Of RepositoryTag)
        Public Overridable Property NoteTags As ICollection(Of NoteTag)

        Public ReadOnly Property PasswordCount As Integer
            Get
                Return PasswordTags.Count
            End Get
        End Property
        Public ReadOnly Property BookCount As Integer
            Get
                Return BookTags.Count
            End Get
        End Property
        Public ReadOnly Property ArticleCount As Integer
            Get
                Return ArticleTags.Count
            End Get
        End Property
        Public ReadOnly Property ActivityCount As Integer
            Get
                Return ActivityTags.Count
            End Get
        End Property
        Public ReadOnly Property RepositoryCount As Integer
            Get
                Return RepositoryTags.Count
            End Get
        End Property
        Public ReadOnly Property NoteCount As Integer
            Get
                Return NoteTags.Count
            End Get
        End Property

        Public Overrides Function GetCaption() As String
            Return String.Format("{0} with Name '{1}'", NameOf(Tag), Name)
        End Function
    End Class
End Namespace
